torch/csrc/utils/throughput_benchmark-inl.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+#include <random>
+#include <thread>
+
+#include <torch/csrc/utils/pybind.h>
+#include <torch/csrc/jit/pybind_utils.h>
+
+namespace torch {
+namespace throughput_benchmark {
+namespace detail {
+
+template <class Input, class Output, class Model>
+BenchmarkExecutionStats BenchmarkHelper<Input, Output, Model>::benchmark(
+    const BenchmarkConfig& config) const {
+  CHECK(initialized_);
+  TORCH_CHECK(
+      config.num_worker_threads == 1,
+      "Only parallelization by callers is supported");
+
+  // We pre-generate inputs here for each of the threads. This allows us to
+  // safely move inputs out for each of the threads independently and thus avoid
+  // overhead from the benchmark runner itself
+  std::vector<std::vector<Input>> thread_inputs(config.num_calling_threads);
+  std::vector<size_t> input_iters(config.num_calling_threads);
+  {
+    std::random_device seeder;
+    std::mt19937 engine(seeder());
+    TORCH_CHECK(
+        !inputs_.empty(),
+        "Please provide benchmark inptus."
+        "Did you forget to call add_input()? ");
+    std::uniform_int_distribution<int> dist(0, inputs_.size() - 1);
+
+    for (int thread_id = 0; thread_id < config.num_calling_threads;
+         ++thread_id) {
+      // Just in case we generate num_iters inputs for each of the threads
+      // This was if one thread does all the work we will be fine
+      for (int i = 0; i < config.num_iters + config.num_warmup_iters; ++i) {
+        thread_inputs[thread_id].push_back(cloneInput(inputs_[dist(engine)]));
+      }
+      input_iters[thread_id] = 0;
+    }
+  }
+
+  std::mutex m;
+  std::condition_variable worker_main_cv;
+  std::condition_variable main_worker_cv;
+  // TODO: add GUARDED_BY once it is available
+  int64_t initialized{0};
+  int64_t finished{0};
+  bool start{false};
+  std::atomic<int64_t> num_forwards{0};
+  std::vector<std::thread> callers;
+
+  // Children need the GIL
+  AutoNoGIL no_gil_guard;
```

Comment:
Are the AutoNoGIL and "// Children need the GIL" still needed?

Reply:
They are still needed. But as Dima pointed out, this indeed can be moved higher up in the stack